### ISSUE TITLE
KDP offers tweaks to CSS for Kindle

### ISSUE THREAD

**Issue by gregturn** [asciidoctor/asciidoctor-epub3#390]
I recently had a correspondence with KDP, inspecting my .epub

```
Greetings from KDP.

I understand that you are looking for an update on your book <redacted>.

It looks like our technical teams have responded. They have stated that there is something wrong with your CSS file that needs to be fixed. I have included their discovery below:

Code snippet: // epub3-css3-only.css

Before the fix:

//Line number - 83
h1.chapter-title .subtitle {
    margin-top: -0.2em;
    margin-bottom: 0.3em; /* compensate for reduced line height */
 }

//Line number - 89
h1.chapter-title .subtitle > b {
    float: left;
    display: inline-block;
    margin-bottom: -0.3em; /* reduce the line height */
    padding-right: 0.2em; /* spacing between words */
 }
After fix:

//Line number - 83
h1.chapter-title .subtitle {
   }

//Line number - 89
h1.chapter-title .subtitle > b {
display: inline-block;
 }

While I am unable to provide any further insight on the situation I hope that this has been helpful. Please let us know if there is anything else we can assist you with.
```

This was in response to my concern about:

```xhtml
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>About the Author</title>
<link rel="stylesheet" type="text/css" href="styles/epub3.css"/>
<link rel="stylesheet" type="text/css" href="styles/epub3-css3-only.css" media="(min-device-width: 0px)"/>
</head>
<body>
<section class="chapter" title="About the Author" epub:type="chapter" id="hacking-with-spring-boot-classic-about">
<header>
<div class="chapter-header">
<h1 class="chapter-title"><small class="subtitle"><b>About</b> <b>the</b> <b>Author</b></small></h1>
</div>
</header>
<p><strong>Greg L. Turnquist</strong> works on the Spring team as a principal developer at VMware. He is a committer to Spring HATEOAS, Spring Data, Spring Boot, R2DBC, and Spring Session for MongoDB. He wrote <em>Hacking with Spring Boot 2.3: Reactive Edition</em> as well as Packt&#8217;s best-selling title, <em>Learning Spring Boot 2.0 2nd Edition</em>. He co-founded the Nashville Java User Group in 2010 and hasn&#8217;t met a Java app (yet) that he doesn&#8217;t like.</p>
<p>Be sure to subscribe to <em>Spring Boot Learning</em>, the YouTube channel where you learn about Spring Boot and having fun doing it at <strong>YouTube.com/GregTurnquist</strong>.</p>
</section>
</body>
</html>
```

Rendering...

<img width="273" alt="Screen Shot 2021-03-03 at 11 58 25 AM" src="https://user-images.githubusercontent.com/311040/109861894-7005e200-7c25-11eb-853c-c16f1332a098.png">

Yet, this bit (which is identical in structure)...

```xhtml
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xml+xhtml; charset=UTF-8"/>
<title>Foreword</title>
<link rel="stylesheet" type="text/css" href="styles/epub3.css"/>
<link rel="stylesheet" type="text/css" href="styles/epub3-css3-only.css" media="(min-device-width: 0px)"/>
</head>
<body>
<section class="chapter" title="Foreword" epub:type="chapter" id="hacking-with-spring-boot-classic-foreword">
<header>
<div class="chapter-header">
<h1 class="chapter-title"><small class="subtitle"><b>Foreword</b></small></h1>
</div>
</header>
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis quis fringilla ex. Curabitur ornare ac lorem ac ullamcorper. Suspendisse blandit accumsan risus a consequat. Quisque sagittis cursus lacus. Curabitur fermentum volutpat malesuada. Aliquam sagittis ex at dolor tincidunt, eu ultrices nisl iaculis. Donec cursus, metus et aliquam pulvinar, mauris quam dignissim dolor, ullamcorper tristique leo elit quis augue. Etiam ultricies ante vitae enim molestie eleifend. Maecenas volutpat ante et ligula ornare mattis. Nam id luctus felis, at faucibus mi. Suspendisse eget erat sem. Aliquam orci elit, vestibulum eu aliquet non, commodo ut urna. Quisque fermentum magna sit amet eros interdum egestas. Mauris ullamcorper rhoncus neque, at consectetur turpis sollicitudin et. Duis semper quam felis, eu imperdiet sapien ultrices non.</p>
<p>Etiam quam dolor, gravida ac tincidunt nec, facilisis eu nunc. Maecenas porta fringilla metus, vitae finibus enim blandit vel. Pellentesque cursus orci eu eros fermentum, non pellentesque leo suscipit. Vivamus eleifend neque nec interdum finibus. Nullam in ante vel urna gravida venenatis at eget augue. Sed tempor quis arcu sed tempus. Suspendisse in felis ultricies, ornare mi et, tempor tortor. Pellentesque quis quam risus. Duis laoreet convallis mollis.</p>
<p>Phasellus pretium ultricies eros quis lacinia. Cras tempus aliquet tellus, id lacinia enim rutrum ac. Nunc consectetur at tortor quis lobortis. Donec dui mauris, venenatis sit amet tellus ac, tempor dictum lectus. Ut dui mauris, tempus et maximus et, blandit non risus. Donec sed lacus non ante ultrices maximus in vel metus. Phasellus nisi lorem, porttitor eu massa sit amet, vehicula dictum lectus. Phasellus ornare felis massa. Nam eu iaculis libero.</p>
<p>&#8201;&#8212;&#8201;<strong>Dr. Dave Syer</strong><br/>
<em>Senior Consulting Engineer and co-founder of Spring Boot</em></p>
</section>
</body>
</html>
```

...renders this...

<img width="267" alt="Screen Shot 2021-03-03 at 11 58 19 AM" src="https://user-images.githubusercontent.com/311040/109861994-8c098380-7c25-11eb-84d5-18e514c238b5.png">

If you want to push a test variant (I have been running `asciidoctor-epub3` through the published Docker container) somewhere, I'd be happy to test it out.

**Comment by slonopotamus** [comment]
@gregturn Do I understand you correctly: we're talking here about feeding .epub (not -kf8.epub) files into Kindle Previewer and looking at how it renders that? Note that modern Kindle Previewer no longer opens .mobi files, so even if you point asciidoctor-epub3 at kindlegen that lives inside Kindle Previewer installation, you cannot olen produced .mobi in Kindle Previewer.

**Comment by gregturn** [comment]
I'm aware that Kindle Previewer 3 doesn't load .mobi files, so I took a shot at .epub file. And it rendered real ugly like that.

Sometime last spring I upgraded my Mac to a version that no longer would run the 32-bit kindlegen, so I started using the Dockerized version of asciidoctor in my build scripts, and didn't realize I was NOT updating.

I checked out the source code of this project to inspect the CSS. Finding everything a little different, I did a `docker pull asciidoctor/docker-asciidoctor` and it pulled a BUNCH of changes.

So I tweaked my build process to roll with the changes, and lo-and-behold, the .epub file NOW looks great! So I'll close this ticket since it looks like whatever issues I had were old data. 

Sorry for the noise!

<img width="605" alt="Screen Shot 2021-03-03 at 2 20 18 PM" src="https://user-images.githubusercontent.com/311040/109867008-a0507f00-7c2b-11eb-8266-2db55337bd97.png">
